### ISSUE TITLE
Thick lines in GLUP

### ISSUE THREAD

**Issue by BrunoLevy** [BrunoLevy/geogram#54]
I miss glLineWidth() ...
Need to write a geometry shader or something to be able to draw thick lines.

**Comment by sebmestrallet** [comment]
Is there any workaround ?
`glupSetMeshWidth()` works well to display mesh wireframes of variable width in Vorpaview, but I didn't manage to use it on custom lines. Is this related ?

**Comment by sebmestrallet** [comment]
I naively added this function in my MeshGfx, by looking at how mesh wireframes were drawn:
```cpp
void CustomMeshGfx::draw_custom_edges() {
      for(/* each group of edges */) {
            glupSetMeshWidth(/* width */);
            glupSetColor4fv(GLUP_FRONT_COLOR, /* color of this group */);
            glupBegin(GLUP_LINES);
            for(/* each edge in this group */) {
                glupPrivateVertex3dv(/* 1st point coordinates */);
                glupPrivateVertex3dv(/* 2nd point coordinates */);
            }
            glupEnd();
      }
}
```
Thank you for the answer, I do not know anything about OpenGL and did not notice the cheat for surfacic primitives!

**Comment by BrunoLevy** [comment]
As a workaround, you may draw the same edges several times, and shift the transform matrix a little bit each time.

**Comment by BrunoLevy** [comment]
- [x] Thick likes support in GLUP_GLSL profile (OpenGL 1.4 and OpenGL 4.4). Works with both immediate and array modes (uses a geometry shader that transforms GL_LINES into little rectangles)
- [ ] Implement in GLUP_ES profile (Macs, Android, WebGL). It will only work in immediate mode (we do not have geometry shaders in this profile).
- [ ] High quality rendering by raytracing capsules

**Comment by BrunoLevy** [comment]
Next release will be published soon, but I have a couple of things to finish in the mesh intersection / boolean ops before shipping it (cannot say exactly when it will be done)

If you are impatient, I keep the main branch quite stable, you can check the status of the [continuous builds](https://brunolevy.github.io/geogram/reports/smoke/) and the [nighty builds](https://brunolevy.github.io/geogram/reports/nightly/) (if it is all green, there are good chances you won't have too many problem updating).

**Comment by BrunoLevy** [comment]
Reopened issue because it does not work very well under Windows as observed by Nicolas Ray.

**Comment by BrunoLevy** [comment]
On Windows, with GLUP_140 and GLUP_440 I get an error: 
`gl_ClipDistance different type between shaders`

**Comment by BrunoLevy** [comment]
Fixed problem under windows:
The problem was that both the vertex and geometry shader were writing to `gl_ClipDistance`, which is supposed to be an error (the error was not detected by the NVidia Linux driver).
Pushed the fix plus some code that save all the generated shaders to files when `gfx:gl_Debug` is set (it is how I figured out what happened, using also glslangValidator to analyze them). 
It is because thick lines are both line primitives and surfacic primitives, and the code that writes `gl_ClipDistance` in the vertex shader does it if it is a line primitive, so I needed also to check that primitive is not GL_THICK_LINES explicitly.
-> pushing a new v1.8.8 release with the fix.
